azhttpclient/options.go

```diff
@@ -3,9 +3,43 @@ package azhttpclient
 import (
 	"github.com/grafana/grafana-azure-sdk-go/azcredentials"
 	"github.com/grafana/grafana-azure-sdk-go/azsettings"
+	"github.com/grafana/grafana-azure-sdk-go/aztokenprovider"
 	sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
 )
 
-func AddAzureAuthentication(opts *sdkhttpclient.Options, settings *azsettings.AzureSettings, credentials azcredentials.AzureCredentials, scopes []string) {
-	opts.Middlewares = append(opts.Middlewares, AzureMiddleware(settings, credentials, scopes))
+type AzureTokenProviderFactory = func(*azsettings.AzureSettings, azcredentials.AzureCredentials) (aztokenprovider.AzureTokenProvider, error)
+
+type AuthOptions struct {
+	settings        *azsettings.AzureSettings
+	scopes          []string
+	customProviders map[string]AzureTokenProviderFactory
```

Comment:
`azsettings.AzureSettings` are global instance-level settings, e.g. in Grafana host they shared across datasources. See `cfg.Azure` [here](https://github.com/grafana/grafana/blob/c81df0dec09dc6f3ba55b9f8faf3342eacda82af/pkg/tsdb/azuremonitor/httpclient.go#L26).

`AuthOptions` is a helper struct for initialization of the middleware, similar to `httpclient.Options` for `httpclient`.

I'm not very concerned about the breaking changes since we know all usages and easily change them. There should not be any breaking changes after that since `AuthOptions` could be extended.

Reply:
I am concerned about the breaking change because it may prevent people from upgrading this dependency in Grafana and this function is currently used there.

Whatever we do, it should be backward compatible (e.g. using a different name than `AddAzureAuthentication`). Update the places in which the old method is used and then we can remove the old method from the SDK.

Re: using a struct or extending the interface. I am not feeling that strongly about that one. Up to you, not sure if @aangelisc has a different opinon.

Reply:
I concur that we should avoid this breaking change if possible. It may not be immediately attainable to update any usages of the `AddAzureAuthentication` function. 

Is there a reason to keep `scopes` separate to `AzureCredentials`? 

I'm fine with the extensibility that will be introduced by using an `interface` for `AuthOptions` (if we choose to keep it).

Reply:
@kostrse any comments here?

Reply:
> Is there a reason to keep scopes separate to AzureCredentials?
Yes, one datasource have only single credentials but it can use them to call different services which require different credentials. So actually different access tokens used for different resources obtained from the same credentials.

Credentials are datasource level. Middleware is httpclient level. Different components may have different http clients to call different resources with different scopes but using the same credentials.
Example is Azure Monitor: [here](https://github.com/grafana/grafana/blob/eb9ef34272c9642706e2caeb04a79eb116a9a8f5/pkg/tsdb/azuremonitor/routes.go#L17) and [here](https://github.com/grafana/grafana/blob/eb9ef34272c9642706e2caeb04a79eb116a9a8f5/pkg/tsdb/azuremonitor/routes.go#L35)

It's possible that even for the same endpoint, different API calls may require different scopes.
E.g. reading schema may require one scope `https://serv.example.org/schema_read` but reading actual data another scope `https://serv.example.org/data_query`. In this case scopes should be set not even on the http client level but also per request.
For this purpose I deliberately removed the `scopes` parameter from the AddAzureAuthentication` function and made it as optional configuration function for `AuthOptions`:

```go
authOpts.Scopes(scopes)
```
This sets default scopes for the http client, the only supported scenario currently.
But it adds possibility to define another way of settings scopes - per request, rather than her http client.

Reply:
Could you give a good example of using interface?

The struct has private fields and public setter methods. So for outside code it's setter only but middleware can read it.
In case of interface, middleware will have to access the data somehow (e.g. [here](https://github.com/kostrse/grafana-azure-sdk-go/blob/05c8dd95b7ef3e88f6dfd68c3e0f0feaa30ef4ef/azhttpclient/middleware.go#L20)) and I don't understand how to implement it cleanly with interface? Expose getter methods to access data? Cast it to a struct?

To me struct looks like a cleaner solution because it setter only for outside code but middleware has access to data. And since it doesn't have any actual logic and just a property bag, there's no need to mock it for testing etc, where interface would actually benefit.

Reply:
Lets implement this as a `struct` for now and if we see any benefits to converting this to an `interface` in the future it can be done then 😊